internal/runner/reattach.go

```diff
@@ -0,0 +1,262 @@
+		return client.SendMsg(m)
+	})
+}
+
+func (c *reattachClient) RecvMsg(m interface{}) error {
+	return c.do(func(client pb.Waypoint_RunnerJobStreamClient) error {
+		return client.RecvMsg(m)
+	})
+}
+
+func (c *reattachClient) do(f func(client pb.Waypoint_RunnerJobStreamClient) error) error {
+	// Shorthand for common attribute access
+	log := c.log
+	r := c.runner
+
+	// Get our configuration state value. We use this so that we can detect
+	// when we've reconnected during failures.
+	stateGen := r.readState(&r.stateConfig)
+
+	// We use this as a way to avoid having unlocks everywhere since we're
+	// using a loop.
+	locked := false
+	defer func() {
+		if locked {
+			c.clientLock.Unlock()
+			locked = false
+		}
+	}()
+
+	for {
+		// If we don't have a lock (our first time through), then we grab it
+		// so we can get our client. If we have a lock (a connection retry)
+		// then we avoid this. We always unlock after this.
+		if !locked {
+			c.clientLock.Lock()
+		}
+
+		// Get our current client
+		client := c.client
+		clientGen := c.clientGen
+		c.clientLock.Unlock()
+		locked = false
+
+		// Do operation
+		err := f(client)
+		if err == nil {
+			// No error means we're done.
+			return nil
+		}
+
+		// We attempt a reconnect on Unavailable and NotFound but otherwise
+		// we return the error as-is.
+		if status.Code(err) != codes.Unavailable &&
+			status.Code(err) != codes.NotFound {
+			return err
+		}
+		log.Warn("server down during job API, will reconnect")
+
+		// Lock so we only ever reconnect once at a time.
+		c.clientLock.Lock()
+		locked = true
+		if c.clientGen != clientGen {
+			// If the client gen changed, we already reconnected.
+			continue
+		}
+
+		// Since this is a disconnect, we have to wait for our
+		// RunnerConfig stream to re-establish. We wait for the config
+		// generation to increment.
+		log.Info("waiting for runner to re-register before reconnecting")
+		if r.waitStateGreater(&r.stateConfig, stateGen) {
+			return status.Error(codes.Internal, "early exit while waiting for reconnect")
+		}
+
+		// Get the new state generation in case we get another disconnect
+		// we can continue retrying and detecting new connections.
+		stateGen = r.readState(&r.stateConfig)
+
+		log.Info("opening job stream")
+		client, err = r.client.RunnerJobStream(context.Background(), grpc.WaitForReady(true))
+		if err != nil {
+			log.Warn("ERROR", "ERROR", err)
+			if status.Code(err) == codes.Unavailable ||
+				status.Code(err) == codes.NotFound {
+				continue
+			}
+
+			return err
+		}
+
+		// Send our request
+		log.Info("sending job request with reattach")
+		if err := client.Send(&pb.RunnerJobStreamRequest{
+			Event: &pb.RunnerJobStreamRequest_Request_{
+				Request: &pb.RunnerJobStreamRequest_Request{
+					RunnerId:      r.id,
+					ReattachJobId: c.jobId,
+				},
+			},
+		}); err != nil {
+			if status.Code(err) == codes.Unavailable ||
+				status.Code(err) == codes.NotFound {
+				continue
```

Comment:
We're abandoning `client` and the top of the loop will use `c.client`, the later I presume is totally dead. Should we be assign `c.client = client` as soon as we get one?

Reply:
That's a good idea. This is a bug due to restrucuring from goto to a loop here.

Reply:
Actually, just used a `goto` to avoid having to execute `f` again, and commented that.